File: android/app/src/main/java/com/viecz/vieczandroid/data/models/Task.kt
package com.viecz.vieczandroid.data.models

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import java.time.Instant

@JsonClass(generateAdapter = true)
data class Task(
    @Json(name = "id") val id: Long,
    @Json(name = "requester_id") val requesterId: Long,
    @Json(name = "tasker_id") val taskerId: Long? = null,
    @Json(name = "category_id") val categoryId: Long,
    @Json(name = "title") val title: String,
    @Json(name = "description") val description: String,
    @Json(name = "price") val price: Long,
    @Json(name = "location") val location: String,
    @Json(name = "latitude") val latitude: Double? = null,
    @Json(name = "longitude") val longitude: Double? = null,
    @Json(name = "status") val status: TaskStatus,
    @Json(name = "deadline") val deadline: String? = null,
    @Json(name = "created_at") val createdAt: String,
    @Json(name = "updated_at") val updatedAt: String,
    @Json(name = "user_has_applied") val userHasApplied: Boolean = false,
    @Json(name = "is_overdue") val isOverdue: Boolean = false
)

enum class TaskStatus {
    @Json(name = "open")
    OPEN,
    @Json(name = "in_progress")
    IN_PROGRESS,
    @Json(name = "completed")
    COMPLETED,
    @Json(name = "cancelled")
    CANCELLED
}

@JsonClass(generateAdapter = true)
data class CreateTaskRequest(
    @Json(name = "title") val title: String,
    @Json(name = "description") val description: String,
    @Json(name = "category_id") val categoryId: Long,
    @Json(name = "price") val price: Long,
    @Json(name = "location") val location: String,
    @Json(name = "latitude") val latitude: Double? = null,
    @Json(name = "longitude") val longitude: Double? = null,
    @Json(name = "deadline") val deadline: String? = null
)

@JsonClass(generateAdapter = true)
data class TasksResponse(
    @Json(name = "data") val data: List<Task>,
    @Json(name = "page") val page: Int,
    @Json(name = "limit") val limit: Int,
    @Json(name = "total") val total: Int
)

/** Client-side overdue check for list views where server doesn't send is_overdue. */
fun Task.computeIsOverdue(): Boolean {
    val dl = deadline ?: return false
    return try {
        Instant.parse(dl).isBefore(Instant.now())
    } catch (_: Exception) {
        false
    }
}

File: server/internal/models/task.go
package models

import (
	"fmt"
	"time"

	"gorm.io/gorm"
)

// TaskStatus represents the status of a task
type TaskStatus string

const (
	TaskStatusOpen       TaskStatus = "open"
	TaskStatusInProgress TaskStatus = "in_progress"
	TaskStatusCompleted  TaskStatus = "completed"
	TaskStatusCancelled  TaskStatus = "cancelled"
)

// Task represents a task in the system
type Task struct {
	ID                int64      `gorm:"primaryKey;autoIncrement" json:"id"`
	RequesterID       int64      `gorm:"not null;index" json:"requester_id"`
	TaskerID          *int64     `gorm:"index" json:"tasker_id,omitempty"`
	CategoryID        int64      `gorm:"not null;index" json:"category_id"`
	Title             string     `gorm:"size:200;not null" json:"title"`
	Description       string     `gorm:"type:text;not null" json:"description"`
	Price             int64      `gorm:"not null" json:"price"`
	Location          string     `gorm:"size:255;not null" json:"location"`
	Latitude          *float64   `gorm:"type:decimal(10,8)" json:"latitude,omitempty"`
	Longitude         *float64   `gorm:"type:decimal(11,8)" json:"longitude,omitempty"`
	Status            TaskStatus `gorm:"type:varchar(20);default:'open';index" json:"status"`
	Deadline          *time.Time `gorm:"column:deadline" json:"deadline,omitempty"`
	CompletedAt       *time.Time `json:"completed_at,omitempty"`
	ImageURLs         []string   `gorm:"type:text[]" json:"image_urls,omitempty"`
	RequesterRatingID *int64     `json:"requester_rating_id,omitempty"`
	TaskerRatingID    *int64     `json:"tasker_rating_id,omitempty"`
	CreatedAt         time.Time  `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt         time.Time  `gorm:"autoUpdateTime" json:"updated_at"`

	// Associations
	Requester User     `gorm:"foreignKey:RequesterID" json:"-"`
	Tasker    *User    `gorm:"foreignKey:TaskerID" json:"-"`
	Category  Category `gorm:"foreignKey:CategoryID" json:"category,omitempty"`
}

// Validate validates the task model
func (t *Task) Validate() error {
	if t.RequesterID == 0 {
		return fmt.Errorf("requester_id is required")
	}

	if t.CategoryID == 0 {
		return fmt.Errorf("category_id is required")
	}

	if t.Title == "" {
		return fmt.Errorf("title is required")
	}

	if len(t.Title) > 200 {
		return fmt.Errorf("title must be less than 200 characters")
	}

	if t.Description == "" {
		return fmt.Errorf("description is required")
	}

	if len(t.Description) > 2000 {
		return fmt.Errorf("description must be less than 2000 characters")
	}

	if t.Price <= 0 {
		return fmt.Errorf("price must be greater than 0")
	}

	if t.Location == "" {
		return fmt.Errorf("location is required")
	}

	if len(t.Location) > 255 {
		return fmt.Errorf("location must be less than 255 characters")
	}

	validStatuses := map[TaskStatus]bool{
		TaskStatusOpen:       true,
		TaskStatusInProgress: true,
		TaskStatusCompleted:  true,
		TaskStatusCancelled:  true,
	}
	if !validStatuses[t.Status] {
		return fmt.Errorf("invalid status: %s", t.Status)
	}

	if len(t.ImageURLs) > 5 {
		return fmt.Errorf("cannot have more than 5 images")
	}

	return nil
}

// BeforeCreate hook is called before creating a task
func (t *Task) BeforeCreate(tx *gorm.DB) error {
	return t.Validate()
}

// BeforeUpdate hook is called before updating a task
func (t *Task) BeforeUpdate(tx *gorm.DB) error {
	return t.Validate()
}

// IsOverdue returns true if the task has a deadline that has passed.
func (t *Task) IsOverdue() bool {
	return t.Deadline != nil && time.Now().After(*t.Deadline)
}
